# What did you expect?

### Learning outcomes

<li> Define an observable, and compute its possible measurement outcomes.

<li> Compute the expectation value of an observable.

Author: [Monit Sharma](https://github.com/MonitSharma)
LinkedIn: [Monit Sharma](https://www.linkedin.com/in/monitsharma/)
Twitter: [@MonitSharma1729](https://twitter.com/MonitSharma1729)
Medium : [MonitSharma](https://medium.com/@_monitsharma)

## Observables 

While obtaining measurement outcome probabilities give us useful information about a qubit's state, we're often interested in other measurable quantities that correspond to something physical, such as its energy.

----

 In quantum mechanics, and consequently quantum computing, physical quantities are related to a special type of object called an **observable**. Observales correspond to *hermitian matrices* whose eigenvalues represent the possible values of the measurement outcome.





*Tip* - A Matrix $B$ is **Hermitian** if $B=B^{\dagger}$. Hermitian matrices have real eigenvalues. 



Expressed mathematically, the expectation value of some observale $B$ for a state $|psi⟩$ is given by 

$$⟨B⟩ = ⟨\psi|B|\psi⟩$$


Namely, we first computer the result of the matrix $B$ applied to $|\psi⟩$ followed by the inner product with $|\psi⟩$

### Code Exercises

Rather than simply measuring outcome probabilities, when solving physical problems, we need to compute a measurable quantity that relates to some property of the system. Such properties are called observables. Mathematically, observables correspond to Hermitian matrices, i.e., matrices $B$ such that $B = B^{\dagger}$ . Each observable has some set of possible measurement outcomes, corresponding to their real eigenvalues. However, since measurement is probabilistic, we usually want to look for the **expectation value**, denoted by $⟨ B ⟩$ , of that physical property

It is very straightforward to compute expectation values in PennyLane. We can simply replace the qml.probs of the previous sections with qml.expval, and specify the observable to be measured. Common choices are qml.PauliX, qml.PauliY, and qml.PauliZ, as many algorithms that solve physical problems require computing the expectation values of the Pauli operators. The possible outcomes of measuring any Pauli-based expectation value are either $1$ or $-1$ , as these are their eigenvalues.

----

*Tip*. These three Pauli operators are unitary and Hermitian. They can therefore be used both as gates in circuits, and considered as observables.

-----

To measure an expectation value in PennyLane, we must specify which observable we are measuring, and which wires it acts on. For example, if we wanted to return a measurement of the PauliZ observable acting on a single qubit, we would write:


```python
@qml.qnode(dev)
def my_circuit():
  #...
  return qml.expval(qml.PauliZ(wires=0))
```



## Codercise I.10.1

Design and run a circuit that performs the following

![](https://codebook.xanadu.ai/pics/exercise_i101.svg)



In [1]:
%pip install pennylane
import pennylane as qml
import numpy as np

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 15.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.3/15.3 MB 37.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.4/46.4 KB 5.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 146.0/146.0 KB 16.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 78.6 MB/s eta 0:00:00


In [2]:
dev = qml.device('default.qubit', wires=1)

@qml.qnode(dev)
def circuit():
    ##################
    # YOUR CODE HERE #
    ##################
    qml.RX(np.pi/4, wires=0)
    qml.Hadamard(wires=0)
    qml.PauliZ(wires=0)

    # IMPLEMENT THE CIRCUIT IN THE PICTURE AND MEASURE PAULI Y

    return qml.expval(qml.PauliY(0))

print(circuit())


-0.7071067811865471


*Tip*- You can also specify a custom observable using qml.Hermitian. For example,



In [ ]:
# Single-qubit Hermitian observable   
O = np.array([[3, 4i], [-4i, 3]])

@qml.qnode(dev)
def circuit():
    # ...
    return qml.expval(qml.Hermitian(O, wires=0)) 




However, one must be careful as specifying an arbitrary observable may make your circuit incompatible with some simulators and hardware.

-----

In the previous sections, we computed measurement outcome probabilities and expectation values analytically. Of course, this is impossible to do on hardware. When we run a circuit and take a measurement, we get a single data point that tells us in which state we observed a qubit for a particular run. Since this process is random, in order to get a clearer picture of the statistics we must perform the experiment many, many times. Each time is typically called a *shot*, or a *sample*. We can *sample* from the output distribution in order to estimate expectation values and measurement outcome probabilities in situations where it isn't possible to do so analytically.


-------

In PennyLane, the number of shots to take during an experiment is specified upon device construction:


```python
dev = qml.device("default.qubit", wires=1, shots=1000)
```



### Codercise I.10.2

In the code below is a list of possible numbers of shots. For each value, initialize a device, then create and QNode that runs circuit from the previous exercise (reproduced below for convenience). What happens to the expectation value as you increase the number of shots?

![](https://codebook.xanadu.ai/pics/exercise_i101.svg)

In [3]:
# An array to store your results
shot_results = []

# Different numbers of shots
shot_values = [100, 1000, 10000, 100000, 1000000]

for shots in shot_values: 
    ##################
    # YOUR CODE HERE #
    ##################
    dev = qml.device('default.qubit', wires=1, shots=shots)
    @qml.qnode(dev)
    def circuit():
        qml.RX(np.pi/4, wires=0)
        qml.Hadamard(wires=0)
        qml.PauliZ(wires=0)
        
        return qml.expval(qml.PauliY(0))

    # CREATE A DEVICE, CREATE A QNODE, AND RUN IT

    # STORE RESULT IN SHOT_RESULTS ARRAY
    shot_results.append(circuit())
    

print(qml.math.unwrap(shot_results))


[-0.68, -0.716, -0.7076, -0.70766, -0.707004]


Unsurprisingly, you'll find that as you increase the number of shots, the results will approach the true value of -0.707.

### Codercise I.10.3

Using the circuit from earlier, replace the `qml.expval` measurement with `qml.sample`. Then, write a function to compute an estimate of the expectation value based on the samples.

In [4]:
dev = qml.device("default.qubit", wires=1, shots=100000)

@qml.qnode(dev)
def circuit():
    qml.RX(np.pi/4, wires=0)
    qml.Hadamard(wires=0)
    qml.PauliZ(wires=0)

    ##################
    # YOUR CODE HERE #
    ##################

    # RETURN THE MEASUREMENT SAMPLES OF THE CORRECT OBSERVABLE

    return qml.sample(qml.PauliY(0))


def compute_expval_from_samples(samples):
    """Compute the expectation value of an observable given a set of 
    sample outputs. You can assume that there are two possible outcomes,
    1 and -1. 
    
    Args: 
        samples (array[float]): 100000 samples representing the results of
            running the above circuit.
        
    Returns:
        float: the expectation value computed based on samples.
    """

    estimated_expval = 0

    ##################
    # YOUR CODE HERE #
    ##################
    estimated_expval = np.mean(samples)

    # USE THE SAMPLES TO ESTIMATE THE EXPECTATION VALUE

    return estimated_expval


samples = circuit()
print(compute_expval_from_samples(samples))


-0.70348


### Codercise I.10.4 
 Using the code below as a starting point, explore how the accuracy of the expectation value depends on the number of shots. For example, if we run 100 experiments with 100 shots each, what are the mean and variance of the distribution of expectation values obtained? How does this variance scale with the number of shots?

In [6]:
def variance_experiment(n_shots):
    """Run an experiment to determine the variance in an expectation
    value computed with a given number of shots.
    
    Args:
        n_shots (int): The number of shots
        
    Returns:
        float: The variance in expectation value we obtain running the 
        circuit 100 times with n_shots shots each.
    """

    # To obtain a variance, we run the circuit multiple times at each shot value.
    n_trials = 100

    ##################
    # YOUR CODE HERE #
    ##################

    # CREATE A DEVICE WITH GIVEN NUMBER OF SHOTS
    
    dev = qml.device("default.qubit", wires=1, shots=n_shots)

    # DECORATE THE CIRCUIT BELOW TO CREATE A QNODE
    @qml.qnode(dev)
    def circuit():
        qml.Hadamard(wires=0)
        return qml.expval(qml.PauliZ(wires=0))

    # RUN THE QNODE N_TRIALS TIMES AND RETURN THE VARIANCE OF THE RESULTS
    results = [circuit().numpy() for _ in range(n_trials)]
    return np.var(results)


def variance_scaling(n_shots):
    """Once you have determined how the variance in expectation value scales
    with the number of shots, complete this function to programmatically
    represent the relationship.
    
    Args:
        n_shots (int): The number of shots
        
    Returns:
        float: The variance in expectation value we expect to see when we run
        an experiment with n_shots shots.
    """

    estimated_variance = 0

    ##################
    # YOUR CODE HERE #
    ##################

    # ESTIMATE THE VARIANCE BASED ON SHOT NUMBER

    return 1/n_shots

# Various numbers of shots; you can change this
shot_vals = [10, 20, 40, 100, 200, 400, 1000, 2000, 4000]

# Used to plot your results
results_experiment = [variance_experiment(shots) for shots in shot_vals]
results_scaling = [variance_scaling(shots) for shots in shot_vals]
plot = plotter(shot_vals, results_experiment, results_scaling)
